src/coreclr/System.Private.CoreLib/src/System/TypedReference.CoreCLR.cs

```diff
@@ -16,6 +19,27 @@ public ref partial struct TypedReference
         private readonly ref byte _value;
         private readonly IntPtr _type;
 
+        private TypedReference(ref byte target, RuntimeType type)
+        {
+            _value = ref target;
+            _type = type.GetUnderlyingNativeHandle();
+        }
+
+        private static unsafe ref byte GetFieldDataReference(object target, RuntimeFieldInfo field)
+        {
+            ByteRef offset = default;
+            GetFieldDataReference(((RtFieldInfo)field).GetFieldDesc(), ObjectHandleOnStack.Create(ref target), ByteRefOnStack.Create(ref offset));
+            Debug.Assert(!Unsafe.IsNullRef(ref offset.Get()));
+            GC.KeepAlive(field);
+            return ref offset.Get();
+        }
+
+        private static int GetFieldOffset(RuntimeFieldInfo field)
+            => RuntimeFieldHandle.GetInstanceFieldOffset((RtFieldInfo)field);
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "TypedReference_GetFieldDataReference")]
+        private static unsafe partial void GetFieldDataReference(IntPtr fieldDesc, ObjectHandleOnStack target, ByteRefOnStack offset);
+
```

Comment:
```suggestion
```
These are reflection internal APIs. They do not belong here - they should live together with other internal reflection APIs (e.g. in RuntimeHandles.cs).

Reply:
I disagree with this approach, but I get the intent. I this we should keep the APIs where they were designed to be used rather than continue to shove them into a giant catch all that we currently have. As API utility increases we should re-designed them appropriately. The discovery issue is real and from an organization perspective I appreciate the approach, but it doesn't seem to be provided a lot of organization in practice. All we wind up with is a myriad of levels of indirection through `Runtime*Handle` types that are either member functions that defer to static or statics that defer to members. It is a mess as is and continuing to put them on the heap isn't really making the situation better.

All that aside, I will put them over there, but all I see we're doing is creating a bucket of functions that no one wants to touch and are far away from were they were designed to be used.

Reply:
I agree that the split between RuntimeHandles, RuntimeType/Field/MethodInfo and InvokeUtils is mess.

I think `GetFieldDataReference` is a step towards moving the top layer of the different Get/SetField QCalls to C#. We have a lot of the required pieces rewritten in C# already.